File: frontend/src/api/home/taskService.js
import axios from "../axiosInstance";

export const createTask = async (taskDto) => {
    const response = await axios.post('/task/', taskDto);
    return response.data;
};
  
export const deleteTask = async (taskId) => {
    const response = await axios.delete(`/task/${taskId}`);
    return response.data;
};

export const updateTask = async (taskId, taskDto) => {
    const response = await axios.put(`/task/${taskId}`, taskDto);
    return response.data;
};

export const getTasks = async (filterParams = {}) => {
    const response = await axios.get('/task/', { params: filterParams });
    return response.data;
};

File: backend/src/main/java/dev/ru/backend/service/TaskService.java
package dev.ru.backend.service;

import dev.ru.backend.model.Task;
import dev.ru.backend.model.User;
import dev.ru.backend.repository.TaskRepository;
import dev.ru.backend.specification.TaskSpecification;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

@Service
public class TaskService {

    private final TaskRepository taskRepository;

    @Autowired
    public TaskService(TaskRepository taskRepository) {
        this.taskRepository = taskRepository;
    }

    public Task createTask(Task task) {
        return taskRepository.save(task);
    }

    public Optional<Task> getTaskById(Long taskId) {
        return taskRepository.findById(taskId);
    }

    public Task updateTask(Task updatedTask) {
        return taskRepository.save(updatedTask);
    }

    public void deleteTask(Long taskId) {
        taskRepository.deleteById(taskId);
    }

    public List<Task> getFilteredTasks(
            User user,
            String title,
            Boolean completed,
            LocalDate finishDate,
            String sortBy,
            String sortDirection
    ) {
        Specification<Task> filters = Specification
                .where(TaskSpecification.hasTitleKeyword(title))
                .and(TaskSpecification.hasStatus(completed))
                .and(TaskSpecification.hasFinishDate(finishDate))
                .and(TaskSpecification.belongsToUser(user));

        Sort.Direction direction = "desc".equalsIgnoreCase(sortDirection) ? Sort.Direction.DESC : Sort.Direction.ASC;
        Sort sort = Sort.by(direction, sortBy);

        return taskRepository.findAll(filters, sort);
    }
}